python/ray/utils.py

```diff
@@ -461,29 +481,53 @@ def get_system_memory():
         return memory_in_bytes
 
 
-def estimate_available_memory():
-    """Return the currently available amount of system memory in bytes.
+def get_used_memory():
+    """Return the currently used system memory in bytes
 
     Returns:
-        The total amount of available memory in bytes. It may be an
-        overestimate if psutil is not installed.
+        The total amount of used memory
     """
+    # Try to accurately figure out the memory usage if we are in a docker
+    # container.
+    docker_usage = None
+    memory_usage_filename = "/sys/fs/cgroup/memory/memory.usage_in_bytes"
+    if os.path.exists(memory_usage_filename):
+        with open(memory_usage_filename, "r") as f:
+            docker_usage = int(f.read())
 
     # Use psutil if it is available.
+    psutil_memory_in_bytes = None
     try:
         import psutil
-        return psutil.virtual_memory().available
+        psutil_memory_in_bytes = psutil.virtual_memory().used
     except ImportError:
         pass
 
-    # Handle Linux.
-    if sys.platform == "linux" or sys.platform == "linux2":
+    if psutil_memory_in_bytes is not None:
+        memory_in_bytes = psutil_memory_in_bytes
+    elif sys.platform == "linux" or sys.platform == "linux2":
+        # Handle Linux.
         bytes_in_kilobyte = 1024
-        return (
-            vmstat("total memory") - vmstat("used memory")) * bytes_in_kilobyte
+        memory_in_bytes = vmstat("used memory") * bytes_in_kilobyte
+    else:
+        # Handle MacOS.
+        memory_in_bytes = sysctl(["sysctl", "hw.memsize"]) - vm_stat("free")
+
+    if docker_usage is not None:
+        return min(docker_usage, memory_in_bytes)
```

Comment:
Seems like it's just a bug if `memory_in_bytes` is less than `docker_usage`, right? Or am I missing something?

Reply:
If so, we can make this function cleaner by just terminating early in the docker case

Reply:
I can't think of a case where that would happen. This was for consistency with the assumption in `get_system_memory` so maybe @robertnishihara knows?

Reply:
Let's print a warning when this happens because it seems like an error case to me